### ISSUE TITLE
Port Trust Metric to Rholang

### ISSUE THREAD

**Issue by kennyrowe** [rchain/bounties#561]
Measure of Completion:  Code Proof of Concept published on Github
Suggested Budget: $5,000

**Comment by kennyrowe** [comment]
@dckc any idea who might like to join this effort?

**Comment by Jake-Gillberg** [comment]
Interested

**Comment by dckc** [comment]
@leithaus I heard you ask about a decentralized version of the trust metric just as today's weekly debrief ended... I guess that could work with public keys and pet names or something... computing the trust metric involves picking a seed, so it's somewhat inherently centralized in that sense.

Are oracles to connect to github and/or discord credentials important? Or shall we just live in the future for this?

What does anybody have in mind for user interfaces to RChain?  Do we have a design for a web3 analog? Maybe just stick to CLI for this? I'm inclined to use the [waterken web_send protocol][1], I think.

[1]: http://waterken.sourceforge.net/web_send/

**Comment by dckc** [comment]
unforgeable names might do fine without any public key crypto. hm.

**Comment by dckc** [comment]
some noodling: https://gist.github.com/dckc/cd9626d556d771f2d710ad7091fed1f9

**Comment by Jake-Gillberg** [comment]
noodles on top of the noodles: https://gist.github.com/Jake-Gillberg/61dbfabbe262f9d86a70d2159a735e9a

**Comment by Jake-Gillberg** [comment]
Realized that you probably need to nest the makePlayer inside of the makeGameBoard contract. You can't just trust that someone will pass you a function that will gather a list of friends without infinite looping or whatever.

**Comment by dckc** [comment]
Good point about never returning... in due course we should be able to represent that as a behavioral type. But for now, yes, having the game board set the player rules makes sense.

**Comment by dckc** [comment]
Our `endorse` method is incomplete without methods to edit and delete endorsements.

**Comment by dckc** [comment]
I wonder how to connect web server apps with rnode.

I like the capper architecture (node.js plus waterken style messaging). It seems compatible with the RhoVM style, but I'm not sure the GRPC interfaces exported by rnode suffice yet. I'm not sure how to do a client side persistent for, for example.

**Comment by David405** [comment]
@dckc I have had similar thoughts too, have you thought of the possibility of RNode working with web3.js or even having a similar JavaScript API? 

By the way, porting the Trust metric to rholang should make it unalterable, what then happens if we intend to make changes to the trust metric?

**Comment by David405** [comment]
Ahhh, I guess the right word should have been "stores persistent records" rather than "unaltered", I now have a better understanding. 

Is Dr SES in use yet or it is still just a proposition? I can't find any useful document as to its usage.

**Comment by David405** [comment]
I am not sure if web3.js supports subscribing to changes on a name, Capper seems interesting also but it is not clear how it can interact with a blockchain node.

**Comment by dckc** [comment]
I'm starting to clean up and document the code I wrote over the last couple days. I moved design notes to [CONTRIBUTING.md](https://github.com/dckc/rchain-dbr/blob/oauth_gateway/CONTRIBUTING.md) and added a _Porting the trust metric to Rholang with node.js and Capper_ section.

**Comment by dckc** [comment]
## trustCertTest deploys signed transaction! (462efb4)

We generate a key for our gateway:

```
<keyPair gateway 1 key: b6251833e2a1...> 'b6251833e2a1210fee3a1ff10ed1ca870ebe65c9832974d90c97250a7d5b935b'
```

Then we suppose these data come from a web form (authenticated with OAuth):

```
    const cert1 = {
	voter: "dckc",
	subject: "bob",
	rating: 2,
	cert_time: clock().toISOString()
    };
```

Using the RSON convention, this becomes:

```
@{"cert_time"}!("2018-07-22T05:37:22.804Z") | @{"rating"}!(2) | @{"subject"}!("bob") | @{"voter"}!("dckc")
```

then we make a call to a **certify** contract with that data and its signature:

```
@{"certify"}!(
  @{"cert_time"}!("2018-07-22T05:37:22.804Z") | @{"rating"}!(2) | @{"subject"}!("bob") | @{"voter"}!("dckc"), "a63e593c8723996483dce9ed820aaeb13ec3ef8b55f1c0851ee608b0b0df8082eafba7cb5ccb7ca7318ad8a92977fe79a6ef00eb6966a621784a99302d1cc806" )
```

and we deploy it using gRPC:

```
05:50:49.203 [grpc-default-executor-193] INFO  c.rchain.casper.MultiParentCasper$ - CASPER: Received Deploy #1532238649198 -- @{"certify"}!(@{"cert_tim...
```

I'm hitting some weird issue with propose (addBlock) via gRPC, but when I `rnode propose`, it likes the new block:

```
05:53:05.134 [repl-io-286] INFO  c.rchain.casper.MultiParentCasper$ - CASPER: New fork-choice tip is block 0d15d9886d....
```

TODO: deploy another contract that checks the signature.

**Comment by kennyrowe** [comment]
:thumbs up:

**Comment by dckc** [comment]
## web UI connected to RChain! (fa1554f)

OAuth parts are still a bit rough.

I'd like to demo this. And/or deploy it to our staging server. Or maybe make a screencast.

For now, how about just a screenshot...

**Comment by David405** [comment]
This is supercool!!!

@dckc is it possible for us to have a zoom session, or better still show it to us during Colab today(perhaps we could deploy it to the staging server there too) or Saturday's office hours? I could use a lot of help from this.

**Comment by dckc** [comment]
Please do try to deploy it to a staging server. I put some effort into
documentation. Let me know where it's not enough.

If collab is during office hours today, no, I'm not available.

**Comment by dckc** [comment]
Anybody got a validating node whose gRPC port I can connect to via an ssh tunnel? Aside from a gRPC server to connect to, I have the app deployed at https://dckc-rchain-dbr.glitch.me/ .

As I mentioned: the OAuth parts are a bit rough. After you log in, replace `?` with `#` in the URL. Then the page should start to do more useful stuff.

Unfortunately, when you hit **Save** it will hang waiting for gRPC. (It comes back after 2 minutes when TCP times out.)

**Comment by jimscarver** [comment]
you don't need a tunnel for grp access to rnode.test.rhobot.net

**Comment by dckc** [comment]
I'm pretty confident that I do, @jimscarver . I'm pretty sure that the gRPC service isn't exported to the public internet. I just tested that I cannot connect to it:

```
Python 2.7.12 (default, Dec  4 2017, 14:50:18)
>>> import socket
>>> c = socket.create_connection(("rnode-test.rhobot.net", 50000))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/socket.py", line 555, in create_connection
    host, port = address
ValueError: too many values to unpack
>>> c = socket.create_connection(("rnode-test.rhobot.net", 50000))
Traceback (most recent call last):
...
socket.error: [Errno 110] Connection timed out

```

The gRPC host defaults to localhost.

https://github.com/rchain/rchain/blob/master/node/src/main/scala/coop/rchain/node/conf.scala#L56-L58

When you run a validating node, you give it a private key. You don't want just anybody on the Internet to be able to connect to it and make use of your key.

**Comment by jimscarver** [comment]
@dckc we only  opened 40000-40405 on the firewall and you are running on 50000. After the command:
ufw allow 50000
the port is now open and accept connections to rnode-test.rhobot.net 50000 from anywhere.

**Comment by jimscarver** [comment]
You can use ufw to control the interface or ip address that is allowed.
https://www.digitalocean.com/community/tutorials/ufw-essentials-common-firewall-rules-and-commands

**Comment by dckc** [comment]
As I showed the code to @JoshOrndorff and @jimscarver we agreed the `rnodeAPI.js` part should go in its own repository / package. So:

 - https://github.com/dckc/rchain-dbr `596f795` rchain-api has moved to its own package
 - https://github.com/dckc/RChain-API `f3748e9`

**Comment by dckc** [comment]
I put together some rough slides on work in this area:

 2a887b8  https://github.com/dckc/RChain-API/tree/talk-reveal/talk

and I managed to make a [pdf version][1]

[1]: https://drive.google.com/file/d/1B8kstgVnpnAKpu1335N0r7paymi57tbv/view?usp=sharing
